Mk 241159 prepopulate locations for edit

mkangia opened the pull request:
http://manage.dimagi.com/default.asp?241159

Pre-populate the locations in the form at render to avoid ajax request causing delay in render.

@benrudolph @esoergel

mkangia committed:
extract get_locations_payload_from_ids from child_locations_for_select2

mkangia committed:
extract get_locations_payload_from_ids from child_locations_for_select2

mkangia committed:
fix exception in case no user on form submit

mkangia commented:
@esoergel i am not very confident if this can cause issues at any other place we are rendering locations. Anything i must test specifically apart from result from tests?

esoergel commented on corehq/apps/locations/views.py:
```diff
@@ -834,6 +834,16 @@ def location_export(request, domain):
+
+
+def base_queryset_for_domain_and_user(domain, user):
+    return SQLLocation.objects.accessible_to_user(domain, user)
```
It just occurred to me that we should make `accessible_to_user` always filter results by domain.  For non-restricted users, this currently doesn't add a domain filter, and I could see things like this confusingly including results from other domains, if that filter isn't explicitly added elsewhere.

esoergel replied:
(I could imagine someone seeing this method without context and assuming that it restricts by domain)

mkangia replied:
@esoergel I think for admin user we would still need to keep `self.all()` since they don't have domain membership and have access to all domain. Will update using that.

esoergel replied:
I don't think we need to worry about admin users, because if code is calling `accessible_to_user(domain, user)`, then they are specifying a domain anyways.  The only time I can think of when we might be intentionally dealing with a cross-domain queryset is in migrations and very advanced type things which presumably wouldn't be using `accessible_to_user` anyways.

One other point - you wrote:
> to be
```python
if user.has_permission(domain, 'access_all_locations'):
    return SQLLocation.active_objects.filter(domain=domain)
```

But it should actually be:
```python
if user.has_permission(domain, 'access_all_locations'):
    return self.filter(domain=domain)
```
This way it will chain properly with other filters that had been applied.

mkangia replied:
Oh yes.. thats right. Admin would get access to locations in the domain only.
Ya! i thought i would give `self` a shot once i actually code this. Any filter would generally be after this so i didn't see chaining as a challenge then. But ya self!

esoergel commented on corehq/apps/users/forms.py:
```diff
@@ -794,13 +799,27 @@ class CommtrackUserForm(forms.Form):
             self.domain = kwargs['domain']
             del kwargs['domain']
+
+        if 'user' in kwargs:
```
Should this just be a required argument?  That might simplify some things, and I can't think of an instance offhand where we wouldn't have `user` available when instantiating the form.

mkangia replied:
That does make sense. Since this form is used for both web and mobile user only and both need the locations pre-populated.

mkangia replied:
Post discussion with Ethan. No changes needed.

esoergel commented:
@mkangia I'd make sure to try out any pages where this form is used, which looks like only a couple.  That's load the page, make sure it prepopulates as expected, save, reload.

mkangia committed:
Filter by domain even for user with access to all location

mkangia committed:
Enable preloading of assigned locations when editing mobile user

mkangia committed:
Fix issue with removing primary location from assigned locations

When the primary location assigned was removed from assigned locations,
the primary location would still get assigned and clear from the view.
This would still get caught at backend but not as expected

mkangia committed:
Alert user if no primary location assigned but locations assigned

mkangia committed:
remove empty strings when fetching locations from list of ids

mkangia committed:
fix form name

mkangia commented:
Closing this and opening a new PR https://github.com/dimagi/commcare-hq/pull/14159 since Sravan took care of preloading locations in https://github.com/dimagi/commcare-hq/pull/14143
